### DISCUSSION TITLE
Module development workflow

### DISCUSSION THREAD

**Discussion by gar1t** [nushell/nushell#10268] [Q&A]
I'm starting on a set of modules/commands and I'm struggling with basic workflow. When I make a change to source I'd like to trivially reload the applicable modules to test the changes from a prompt. I don't mind reloading everything. Not knowing how else to do this I'm starting a new shell and calling `use <mod> *` to pick up the latest changes.

**Comment by ghost** [comment]
I usually use one of the following workflows (depending on what i'm writing)
- `overlay use <mod>` (equivalent to `use <mod> *`) (`overlay hide` is usually optional)
- when working on my config: `nu` (new nested shell, which i close after testing it)
- just run `use …` again without closing anything - it should override it
- write smaller code sections in the REPL and just copy the working code into the file and test the file once
- use a custom vim-keybind, which generates a temporary config.nu and opens a new instance

i also use [a small script](https://github.com/Jan9103/pre-commit-nushell/blob/main/check-ast) to check for errors before even loading it

in case youre into it you can also write a expect script, which auto-tests it

but as someone who has written a lot of nushell scripts a warning: when your script needs modules it might be comfortable to write, but if you're not constantly updating it (and running the same nu version on all of your computers) the maintenance can get annoying really quick - just a few examples for braking changes you'll have to constantly deal with: `let-env` -> `$env`, `get -i $foo` -> `try {get $foo} catch {null}` (and back), removal of `echo`, `parse "^foo"` -> `parse -r "^foo"`, … (just look at the breaking changes section of change log of the last releases)   
and also nu might already be great as a shell, but especially for scripting there are still a lot of (known) bugs and missing features, such as: not being able to load hooks from module, no comments in matches, nu crashing when a function takes 6 arguments, …

**Comment by gar1t** [comment]
Cool thanks for the advice. I'm opting for a test driven approach where I run test files that setup the env each time.

Coming from interpreted languages I'm accustomed to editing a source file and having that change available immediately without additional ceremony.

It would be nice if Nushell had a `reload` feature that re-reloaded custom modules (hand wave here as I don't know the boundary between custom and core, or if its convenient).

**Comment by ghost** [comment]
i hacked together a little something - a reload command dosnt work since the only way to remove code is by using overlays and functions not beeing able to edit envs (and aliases only beeing able to hold a single command)

"installation":
1. drop the following into any nu file (i use `dev.nu`)
2. `touch /tmp/nu_dev.nu`

usage:
1. `use dev.nu` (or however you called it)
2. add all your modules: `dev add FILE LOAD-TYPE` (load-types: `use` = `use foo.nu`, `use_all` = `use foo.nu *`, `env` loads things written in `export-env`)
3. `dev start`
4. to reload just `crtl+d` or `exit`
5. to exit `exit 1` (where you can run `dev add` / `dev remove` again)

```nu
export-env {
  $env.DEV_MODS = []
  "" | save -rf /tmp/nu_dev.nu
}


def types [] {[use use_all env]}


export def-env add [
  module_file: path
  type: string@"types"
] {
  $env.DEV_MODS = ($env.DEV_MODS | append [[($module_file | path expand) $type]])
  rebuild
}


export def-env remove [module_file: path] {
  $env.DEV_MODS = ($env.DEV_MODS | where $it.0 != ($module_file | path expand))
  rebuild
}


def rebuild [] {
  ([
    "export-env {"
      ($env.DEV_MODS | where $it.1 == "env" | each {|i| $"use ($i.0)"})
    "}"
    ($env.DEV_MODS | where $it.1 == "use" | each {|i| $"export use ($i.0)"})
    ($env.DEV_MODS | where $it.1 == "use_all" | each {|i| $"export use ($i.0) *"})
  ] | flatten | str join "\n" | save -rf "/tmp/nu_dev.nu")
}


export def-env start [] {
  while true {
    print "reloaded env - to exit the env use 'exit 1'"
    nu -e "overlay use /tmp/nu_dev.nu"
    if $env.LAST_EXIT_CODE != 0 {return}
  }
}
```
